src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,269 @@
+// Otherwise, returns OK and cached response.
+::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+  if (!lookup.Found()) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  CacheElem* elem = lookup.value();
+
+  // If the cached check response has check errors, then we assume the new
+  // request should fail as well and return the cached check response.
+  // However, after the flush interval, the first check request will be send to
+  // the server to refresh the check response. Other check requests still fail
+  // with cached check response.
+  // If the cached check response is a pass, then we assume the new request
+  // should pass as well and return the cached response directly, besides
+  // updating the quota info to be the same as requested. The requested tokens
+  // are aggregated until flushed.
+  // More details can be found in design doc go/simple-chemist-client.
+  if (elem->check_response().allocate_errors_size() > 0) {
+    if (ShouldFlush(*elem)) {
+      // Pretend that we did not find, so we can force it into a check request
+      // to the server.
+      //
+      // Setting last check to now to block more check requests to Chemist.
+      elem->set_last_check_time(SimpleCycleTimer::Now());
+      // By returning NO_FOUND, caller will send request to server.
+      return Status(Code::NOT_FOUND, "");
+    } else {
+      // Use cached response.
+      *response = elem->check_response();
+      return Status::OK;
+    }
+  } else {
+    elem->Aggregate(request, metric_kinds_.get());
+
+    if (ShouldFlush(*elem)) {
+      if (elem->is_flushing()) {
+        GOOGLE_LOG(WARNING) << "Last refresh request was not completed yet.";
+      }
+
+      elem->set_is_flushing(true);
+
+      // Setting last check to now to block more check requests to Chemist.
+      elem->set_last_check_time(SimpleCycleTimer::Now());
+
+      // By returning NO_FOUND, caller will send request to server.
+      return Status(Code::NOT_FOUND, "");
```

Comment:
This is not good.   We need to call with aggregated metrics.   This is the different between Check() and AllocateQuota()